pkg/pool-stable-phantom/contracts/StablePhantomPool.sol

```diff
@@ -0,0 +1,466 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/EnumerableMap.sol";
+
+import "@balancer-labs/v2-pool-stable/contracts/StablePool.sol";
+import "@balancer-labs/v2-pool-utils/contracts/rates/PriceRateCache.sol";
+import "@balancer-labs/v2-pool-utils/contracts/interfaces/IRateProvider.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/math/FixedPoint.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/ERC20Helpers.sol";
+import "@balancer-labs/v2-solidity-utils/contracts/helpers/BalancerErrors.sol";
+
+contract StablePhantomPool is StablePool {
+    using WordCodec for bytes32;
+    using FixedPoint for uint256;
+    using PriceRateCache for bytes32;
+    using EnumerableMap for EnumerableMap.IERC20ToBytes32Map;
+
+    uint256 private constant _MIN_TOKENS = 2;
+    uint256 private constant _MINIMUM_BPT = 0;
+    uint256 private constant _MAX_TOKEN_BALANCE = 2**(112) - 1;
+
+    uint256 private immutable _bptIndex;
+
+    // Price rate caches are used to avoid querying the price rate for a token every time we need to work with it.
+    // Data is stored with the following structure:
+    //
+    // [   expires   | duration | price rate value ]
+    // [   uint64    |  uint64  |      uint128     ]
+
+    mapping(IERC20 => bytes32) private _priceRateCaches;
+    mapping(IERC20 => IRateProvider) private _rateProviders;
+
+    uint256 private constant _PRICE_RATE_CACHE_VALUE_OFFSET = 0;
+    uint256 private constant _PRICE_RATE_CACHE_DURATION_OFFSET = 128;
+    uint256 private constant _PRICE_RATE_CACHE_EXPIRES_OFFSET = 128 + 64;
+
+    event PriceRateCacheUpdated(IERC20 indexed token, uint256 rate);
+    event PriceRateProviderSet(IERC20 indexed token, IRateProvider indexed provider, uint256 cacheDuration);
+
+    // The constructor arguments are received in a struct to work around stack-too-deep issues
+    struct NewPoolParams {
+        IVault vault;
+        string name;
+        string symbol;
+        IERC20[] tokens;
+        IRateProvider[] rateProviders;
+        uint256[] priceRateCacheDurations;
+        uint256 amplificationParameter;
+        uint256 swapFeePercentage;
+        uint256 pauseWindowDuration;
+        uint256 bufferPeriodDuration;
+        address owner;
+    }
+
+    constructor(NewPoolParams memory params)
+        StablePool(
+            params.vault,
+            params.name,
+            params.symbol,
+            _insertSorted(params.tokens, IERC20(this)),
```

Comment:
Small note. Stable pool can handle 5 tokens + bpt because bpt are not part of invariant math. We are limiting it to 4 tokens + bpt. I think it is ok for our current use of this pool.